pkg/apply/prune/prune.go

```diff
@@ -83,114 +83,174 @@ type Options struct {
 			continue
 		}
-		// If this cluster object is not also a currently applied
-		// object, then it has been omitted--prune it. If the cluster
-		// object is part of the local apply set, skip it.
-		uid := string(metadata.GetUID())
-		klog.V(7).Infof("prune previously applied object UID: %s", uid)
+		// Do not prune objects that are in set of currently applied objects.
+		uid := string(obj.GetUID())
 		if currentUIDs.Has(uid) {
-			klog.V(7).Infof("prune object in current apply; do not prune: %s", uid)
+			klog.V(5).Infof("prune object in current apply; do not prune: %s", uid)
 			continue
 		}
 		// Handle lifecycle directive preventing deletion.
 		if !canPrune(localInv, obj, o.InventoryPolicy, uid) {
-			taskContext.EventChannel() <- createPruneEvent(clusterObj, obj, event.PruneSkipped)
-			localIds = append(localIds, clusterObj)
+			klog.V(4).Infof("skip prune for lifecycle directive %s/%s", pruneObj.Namespace, pruneObj.Name)
+			taskContext.EventChannel() <- createPruneEvent(pruneObj, obj, event.PruneSkipped)
+			pruneFailures = append(pruneFailures, pruneObj)
 			continue
 		}
 		// If regular pruning (not destroying), skip deleting namespace containing
 		// currently applied objects.
 		if !po.Destroy {
-			if clusterObj.GroupKind == object.CoreV1Namespace.GroupKind() &&
-				localNamespaces.Has(clusterObj.Name) {
-				klog.V(4).Infof("skip pruning namespace: %s", clusterObj.Name)
-				taskContext.EventChannel() <- createPruneEvent(clusterObj, obj, event.PruneSkipped)
-				localIds = append(localIds, clusterObj)
-				taskContext.CaptureResourceFailure(clusterObj)
+			if pruneObj.GroupKind == object.CoreV1Namespace.GroupKind() &&
+				localNamespaces.Has(pruneObj.Name) {
+				klog.V(4).Infof("skip pruning namespace: %s", pruneObj.Name)
+				taskContext.EventChannel() <- createPruneEvent(pruneObj, obj, event.PruneSkipped)
+				pruneFailures = append(pruneFailures, pruneObj)
+				taskContext.CaptureResourceFailure(pruneObj)
 				continue
 			}
 		}
 		if !o.DryRunStrategy.ClientOrServerDryRun() {
-			klog.V(4).Infof("prune object delete: %s/%s", clusterObj.Namespace, clusterObj.Name)
-			err = namespacedClient.Delete(context.TODO(), clusterObj.Name, metav1.DeleteOptions{})
+			klog.V(4).Infof("prune object delete: %s/%s", pruneObj.Namespace, pruneObj.Name)
+			namespacedClient, err := po.namespacedClient(pruneObj)
 			if err != nil {
-				taskContext.EventChannel() <- createPruneFailedEvent(clusterObj, err)
-				localIds = append(localIds, clusterObj)
-				taskContext.CaptureResourceFailure(clusterObj)
+				if klog.V(4) {
+					klog.Errorf("prune failed for %s/%s (%s)", pruneObj.Namespace, pruneObj.Name, err)
+				}
+				taskContext.EventChannel() <- createPruneFailedEvent(pruneObj, err)
+				pruneFailures = append(pruneFailures, pruneObj)
+				taskContext.CaptureResourceFailure(pruneObj)
 				continue
 			}
+			err = namespacedClient.Delete(context.TODO(), pruneObj.Name, metav1.DeleteOptions{})
+			if err != nil {
```

Comment:
The error handling here is the same as the one after `namespacedClient, err := po.namespacedClient(pruneObj)`. We can combine the getting client and deleting the object into one function, like what you have done for `getObject`. The new function can be `deleteObject`.